geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/writer/tx/AccumuloAtomicIndexWriter.scala

```diff
@@ -0,0 +1,208 @@
+    mutate(feature.feature.getID, buildMutations(values, AppendBuilder.apply))
+
+  override protected def delete(feature: WritableFeature, values: Array[RowKeyValue[_]]): Unit =
+    mutate(feature.feature.getID, buildMutations(values, DeleteBuilder.apply))
+
+  override protected def update(
+      feature: WritableFeature,
+      values: Array[RowKeyValue[_]],
+      previous: WritableFeature,
+      previousValues: Array[RowKeyValue[_]]): Unit = {
+    val mutations = Array.ofDim[Seq[MutationBuilder]](values.length)
+    // note: these are temporary conditions - we update them below
+    val updates = buildMutations(values, AppendBuilder.apply)
+    val deletes = buildMutations(previousValues, DeleteBuilder.apply)
+    var i = 0
+    while (i < values.length) {
+      val dels = ArrayBuffer(deletes(i): _*)
+      val puts = updates(i).map { mutations =>
+        // find any previous values that will be updated
+        val d = dels.indexWhere(p => java.util.Arrays.equals(p.row, mutations.row))
+        if (d == -1) { mutations } else {
+          // note: side-effect
+          val toUpdate = dels.remove(d)
+          // any previous values that are overwritten are added as conditions on the put
+          // any previous values that aren't overwritten need to be deleted
+          val remaining = toUpdate.kvs.filterNot(kv => mutations.kvs.exists(_.equalKey(kv)))
+          if (remaining.nonEmpty) {
+            dels.append(toUpdate.copy(kvs = remaining))
+          }
+          UpdateBuilder(mutations.row, mutations.kvs, toUpdate.kvs)
+        }
+      }
+      // note: order here is important, we need puts to operate before deletes in order for conditions to succeed
+      mutations(i) = puts ++ dels
+      i += 1
+    }
+    mutate(feature.feature.getID, mutations)
+  }
+
+  /**
+   * Apply the mutations
+   *
+   * @param id feature id
+   * @param mutations mutations to apply
+   */
+  @throws[ConditionalWriteException]
+  private def mutate(id: String, mutations: Array[Seq[MutationBuilder]]): Unit = {
+    val errors = ArrayBuffer.empty[ConditionalWriteStatus]
+    val successes = ArrayBuffer.empty[(Int, MutationBuilder)]
+    var i = 0
+    while (i < mutations.length) {
+      // note: mutations need to be applied in order so that conditions are correct
+      mutations(i).foreach { m =>
+        val status = writers(i).write(m.apply()).getStatus
+        if (status == ConditionalWriter.Status.ACCEPTED ||
+            (status == ConditionalWriter.Status.UNKNOWN && verifyWrite(tables(i), m))) {
+          successes += i -> m
+        } else {
+          errors += ConditionalWriteStatus(indices(i), m.name, status)
+        }
+      }
+      i += 1
+    }
+    if (errors.nonEmpty) {
+      // revert any writes we made
+      successes.foreach { case (i, m) =>
+        // note: if these are rejected, some other update has come through and we don't want to overwrite it
+        writers(i).write(m.invert())
```

Comment:
Does the order in which we revert matter?

Reply:
it shouldn't matter since they're all independent tables/rows